rss-bridge: add config option

Mynacol opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->
This allows managing rss-bridge's config with nix. Config options are listed at
https://rss-bridge.github.io/rss-bridge/For_Hosts/Custom_Configuration.html

This change should be backwards compatible, only setting the `/var/lib/rss-bridge/config.ini.php` symlink if actually using the new config option (similar to the whitelist option).

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

-------------
(first contribution, open for improvements)

h7x4 commented on nixos/modules/services/web-apps/rss-bridge.nix:
```diff
@@ -7,6 +7,11 @@ let
+
+  configFile = pkgs.writeText "rss-bridge_config.ini" ''
+    <?php exit; ?>
+    ${lib.generators.toINI {} cfg.config}
```
Maybe you could make use of `pkgs.formats.ini` here? That would also let you be more specific with the format for the option.

```nix
# somewhere
format = pkgs.formats.ini { };

# here
configFile = pkgs.writeText "rss-bridge_config.ini" ''
<?php exit; ?>
${format.generate cfg.config}
'';
```

h7x4 commented on nixos/modules/services/web-apps/rss-bridge.nix:
```diff
@@ -72,6 +77,20 @@ in
       };
+
+      config = mkOption {
+        type = types.attrs;
```
(ref previous comment)

Here, you could do

```nix
type = format.type;
```

Quantenzitrone commented:
Having the config file in `/etc` would be nice, but i don't see it as necessary, I'd rather have this merged soon, since the current way of whitelisting doesn't work anymore.

Mynacol committed:
rss-bridge: add config option

This allows managing rss-bridge's config with nix.
It leverages the environment variable way of setting the config options,
introduced quite [some time ago](https://github.com/RSS-Bridge/rss-bridge/pull/2100)
It is the only existing way to set config options independent of the
document root, and upstream is [hesitant](https://github.com/RSS-Bridge/rss-bridge/pull/3842)
to change the config loading methods.

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

Mynacol committed:
rss-bridge: Move whitelist option to general config

Prefer setting the whitelisted bridges through the generic configuration
method. Removes the need for a whitelist.txt file.

Preserves backwards compatibility by taking the same values and
essentially just renaming the config option.

Mynacol committed:
rss-bridge: Integrate filecache path with config

Preserve the default value for the filecache path, but also allow
modifying it, adapting the tmpfiles rule to create the directory with
the right permissions.

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

SuperSandro2000 commented on nixos/modules/services/web-apps/rss-bridge.nix:
```diff
@@ -94,8 +106,7 @@ in
     systemd.tmpfiles.rules = [
-      "d '${cfg.dataDir}/cache' 0750 ${cfg.user} ${cfg.group} - -"
-      (mkIf (cfg.whitelist != []) "L+ ${cfg.dataDir}/whitelist.txt - - - - ${whitelist}")
+      "d '${configAttr.FileCache.path}' 0750 ${cfg.user} ${cfg.group} - -"
```
Can we switch this to the new tmpfiles syntax?

SuperSandro2000 commented on nixos/modules/services/web-apps/rss-bridge.nix:
```diff
@@ -56,20 +64,24 @@ in
-          List of bridges to be whitelisted.
-          If the list is empty, rss-bridge will use whitelist.default.txt.
-          Use `[ "*" ]` to whitelist all.
+          Attribute set of arbitrary config options. Please consult the documentation at the [wiki](https://rss-bridge.github.io/rss-bridge/For_Hosts/Custom_Configuration.html) and [sample config](https://github.com/RSS-Bridge/rss-bridge/blob/master/config.default.ini.php) to see a list of available options.
```
```suggestion
          Attribute set of arbitrary config options.
          Please consult the documentation at the [wiki](https://rss-bridge.github.io/rss-bridge/For_Hosts/Custom_Configuration.html)
          and [sample config](https://github.com/RSS-Bridge/rss-bridge/blob/master/config.default.ini.php) to see a list of available options.
```

SuperSandro2000 reviewed (commented):
the other parts are a bit to complicated for me right now, maybe I look at this again at a later point

Mynacol committed:
rss-bridge: Use new tmpfiles syntax

SuperSandro2000 commented on nixos/modules/services/web-apps/rss-bridge.nix:
```diff
@@ -56,20 +64,26 @@ in
-        type = types.listOf types.str;
-        default = [];
+      config = mkOption {
+        type = types.attrsOf (types.attrsOf (types.oneOf [ types.bool types.int types.str (types.listOf types.str) ]));
```
```suggestion
        type = with types; attrsOf (attrsOf (oneOf [ bool int str (listOf str) ]));
```
nit

SuperSandro2000 commented on nixos/modules/services/web-apps/rss-bridge.nix:
```diff
@@ -5,10 +5,18 @@ let
 
-  whitelist = pkgs.writeText "rss-bridge_whitelist.txt"
-    (concatStringsSep "\n" cfg.whitelist);
+  configAttr = { FileCache.path = "${cfg.dataDir}/cache/"; } // cfg.config;
```
```suggestion
  configAttr = lib.recursiveUpdate { FileCache.path = "${cfg.dataDir}/cache/"; } cfg.config;
```
otherwise specifying just FileCache will delete the path

SuperSandro2000 commented on nixos/modules/services/web-apps/rss-bridge.nix:
```diff
@@ -5,10 +5,18 @@ let
-  whitelist = pkgs.writeText "rss-bridge_whitelist.txt"
-    (concatStringsSep "\n" cfg.whitelist);
+  configAttr = { FileCache.path = "${cfg.dataDir}/cache/"; } // cfg.config;
+  cfgHalf = lib.attrsets.mapAttrsRecursive (path: value: let
```
```suggestion
  cfgHalf = lib.mapAttrsRecursive (path: value: let
```

SuperSandro2000 commented on nixos/modules/services/web-apps/rss-bridge.nix:
```diff
@@ -5,10 +5,18 @@ let
+  configAttr = { FileCache.path = "${cfg.dataDir}/cache/"; } // cfg.config;
+  cfgHalf = lib.attrsets.mapAttrsRecursive (path: value: let
+    envName = lib.toUpper ("RSSBRIDGE_" + lib.concatStringsSep "_" path);
+    envValue = if lib.isList value then lib.concatStringsSep "," value else if lib.isBool value then lib.boolToString value else builtins.toString value;
```
```suggestion
    envValue = if lib.isList value then
      lib.concatStringsSep "," value
    else if lib.isBool value then
      lib.boolToString value
    else
      toString value;
```
Lets make this a bit more readable

SuperSandro2000 commented on nixos/modules/services/web-apps/rss-bridge.nix:
```diff
@@ -5,10 +5,18 @@ let
+    envName = lib.toUpper ("RSSBRIDGE_" + lib.concatStringsSep "_" path);
+    envValue = if lib.isList value then lib.concatStringsSep "," value else if lib.isBool value then lib.boolToString value else builtins.toString value;
+  in "fastcgi_param \"${envName}\" \"${envValue}\";") configAttr;
+  cfgEnv = lib.concatStringsSep "\n" (lib.attrsets.collect lib.isString cfgHalf);
```
```suggestion
  cfgEnv = lib.concatStringsSep "\n" (lib.collect lib.isString cfgHalf);
```

SuperSandro2000 reviewed (commented):
I've mostly added some formatting tips to make this easier to understand for me. Can you squash the review commits into one or two meaningful ones?  Then I want to cherry-pick and test this on my instance and if that continues to work, we can merge this.

SuperSandro2000 reviewed (approved):
my config continues to work like before but it is also pretty simple and just consists of a whitelist
